bin/propolis-standalone/src/snapshot.rs

```diff
@@ -55,22 +49,40 @@ struct SnapshotDevice {
 struct SnapshotDevicePayload {
     pub kind: String,
     pub version: u32,
-    pub data: Vec<u8>,
+    pub data: String,
 }
 
+const DEVICE_DIR: &str = "devices";
+const MEMORY_DIR: &str = "memory";
+const CONFIG_NAME: &str = "config.toml";
+const GLOBAL_NAME: &str = "global.json";
+
 /// Save a snapshot of the current state of the given instance to disk.
-pub(crate) async fn save(
+pub(crate) fn save(
     inst: &propolis::Instance,
     config: &Config,
     log: &slog::Logger,
 ) -> anyhow::Result<()> {
-    let now = std::time::SystemTime::now()
-        .duration_since(std::time::UNIX_EPOCH)?
-        .as_millis();
-    let snapshot = format!("{}-{}.bin", config.main.name, now);
+    let now = SystemTime::now().duration_since(UNIX_EPOCH)?;
+    let snapshot = format!("{}-{}.tar", config.main.name, now.as_millis());
 
     info!(log, "saving snapshot of VM to {}", snapshot);
 
+    let file =
+        File::create(&snapshot).context("Failed to create snapshot file")?;
+    let mut builder = TarBuilder::new(file);
+    let mut header = tar::Header::new_gnu();
+    header.set_mode(0o444);
+    header.set_mtime(now.as_secs());
+
+    info!(log, "Serializing VM config");
+    {
+        let config_bytes = toml::to_string(config)?.into_bytes();
+        header.set_size(config_bytes.len() as u64);
+        header.set_cksum();
+        builder.append_data(&mut header, CONFIG_NAME, &config_bytes[..])?;
```

Comment:
You can skip the `header.set_cksum()` call since [`Builder::append_data`](https://docs.rs/tar/latest/tar/struct.Builder.html#method.append_data) will do it for you:
>The checksum for the header will be automatically updated via the `set_cksum` method after setting the path.